pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/PersistentTransactionBuffer.java

```diff
@@ -0,0 +1,308 @@
+        public long getSequenceId() {
+            return this.sequenceId;
+        }
+    }
+
+    @Builder
+    private static final class Marker implements Serializable {
+        TxnID txnID;
+        TxnStatus status;
+
+        public byte[] serialize() {
+            return SerializationUtils.serialize(this);
+        }
+    }
+
+    public PersistentTransactionBuffer(String topic, ManagedLedger ledger, BrokerService brokerService)
+        throws BrokerServiceException.NamingException, ManagedLedgerException {
+        super(topic, ledger, brokerService);
+        this.txnCursor = new TransactionCursorImpl();
+        this.retentionCursor = ledger.newNonDurableCursor(PositionImpl.earliest);
+    }
+
+    @Override
+    public CompletableFuture<TransactionMeta> getTransactionMeta(TxnID txnID) {
+        return txnCursor.getTxnMeta(txnID, false);
+    }
+
+    @Override
+    public CompletableFuture<Void> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
+        CompletableFuture<Void> appendFuture = new CompletableFuture<>();
+
+        publishMessage(buffer, new TxnCtx(sequenceId) {
+            @Override
+            public void completed(Exception e, long ledgerId, long entryId) {
+                if (e != null) {
+                    appendFuture.completeExceptionally(e);
+                    return;
+                }
+
+                txnCursor.getTxnMeta(txnId, true).thenCompose(meta -> {
+                    meta.appendEntry(sequenceId, PositionImpl.get(ledgerId, entryId));
+                    appendFuture.complete(null);
+                    return null;
+                }).exceptionally(metaNotFound -> {
+                    appendFuture.completeExceptionally(metaNotFound);
+                    return null;
+                });
+            }
+        });
+
+        return appendFuture;
+    }
+
+    @Override
+    public CompletableFuture<TransactionBufferReader> openTransactionBufferReader(TxnID txnID, long startSequenceId) {
+        CompletableFuture<TransactionBufferReader> readerFuture = new CompletableFuture<>();
+
+        txnCursor.getTxnMeta(txnID, false).thenCompose(meta -> {
+            try {
+                PersistentTransactionBufferReader reader = new PersistentTransactionBufferReader(meta, ledger);
+                readerFuture.complete(reader);
+            } catch (ManagedLedgerException | TransactionNotSealedException e) {
+                readerFuture.completeExceptionally(e);
+            }
+            return null;
+        }).exceptionally(e -> {
+            readerFuture.completeExceptionally(e);
+            return null;
+        });
+
+        return readerFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> commitTxn(TxnID txnID, long committedAtLedgerId, long committedAtEntryId) {
+        CompletableFuture<Void> commitFuture = new CompletableFuture<>();
+
+        Marker commitMarker = Marker.builder().txnID(txnID).status(TxnStatus.COMMITTED).build();
```

Comment:
How can you differentiate the normal messages from the marker messages?

You need to publish a normal pulsar message here with marker_type.

1) You need to introduce transaction related marker types: pulsar-common/src/main/proto/PulsarMarkers.proto
2) Defined the commit and abort marker entries: pulsar-common/src/main/proto/PulsarMarkers.proto